Add static builds for various platforms

jagerman opened the pull request:
Adds static bundles (that get uploaded via CI) for various platforms:
- Linux amd64/i386/arm64/armhf
- Windows x64/i686
- macOS (single multiarch lib for both x86_64 and arm64)
- Android (x86_64, armeabi-v7a, arm64-v8a, and x86)
- iOS (arm64 for devices + multiarch lib for x86_64 arm64 simulators)

jagerman committed:
Add static bundle builds for all platforms

jagerman committed:
Rename contrib -> utils

Nothing in here is actually a non-core contribution, so the contrib name
seems wrong.

majestrate commented:
i like the way things that dont belong in contrib arent in contrib and want to add that to lokinet.

majestrate commented on utils/ios.sh:
```diff
@@ -0,0 +1,79 @@
+set -e
+
+if ! [ -f LICENSE ] || ! [ -d include/session ]; then
+    echo "You need to run this as ./utils/ios.sh from the top-level libsession-util project directory"
```
should this write to stderr as the one below does ?

majestrate commented on utils/ios.sh:
```diff
@@ -0,0 +1,79 @@
+    echo "Building ${i[0]} for $platform in $build"
+
+    ./utils/static-bundle.sh "$build" "" \
+        -DCMAKE_TOOLCHAIN_FILE=../../external/ios-cmake/ios.toolchain.cmake \
```
use absolute path here as having it be a relative the way it is hard to understand.

jagerman replied:
Relative is easier when you are running the same commands manually while looking at the script (e.g. to debug something).

majestrate commented on utils/ios.sh:
```diff
@@ -0,0 +1,79 @@
+    -output "$pkg_dir/libsession-util.xcframework"
+
+# Copy the headers over
+cp -rv include/session $pkg_dir/libsession-util.xcframework
```
should those two paths to the xcframework be escaped the same?

majestrate commented on utils/ios.sh:
```diff
@@ -0,0 +1,79 @@
+cp -rv include/session $pkg_dir/libsession-util.xcframework
+
+# The 'module.modulemap' is needed for XCode to be able to find the headers
+modmap="$pkg_dir/libsession-util.xcframework/module.modulemap"
```
i feel like cmake should make this file honestly

jagerman replied:
This is only something ios cares about so not worth the bother.

jagerman committed:
Remove obsolete closed group implementation

Will resurrect this and adapt to new API when implementing it, but for
now it's just breaking things.

jagerman committed:
Don't tell xcode about C++

It breaks (by trying to load them all as C) if we do because Apple's
xcode toolchain is a disaster.

jagerman committed:
Add oxenc headers

yarn, in particular, is apparently unable to use submodules because the
bug for supporting them has only been open for 6 years.

jagerman committed:
Static bundle updates

- Disable LTO for static bundles
- set android host via android cmake vars

jagerman committed:
Experiment: don't merge android .a's

jagerman committed:
Fixups for building as a subproject

jagerman committed:
Android toolchain mess

jagerman committed:
Rename libs

libcrypto.so -> libsession-util-crypto.so
libconfig.so -> libsession-util-config.so
libsodium.so -> libsodium-internal.so

jagerman committed:
Properly rename the libsodium-internal in the fork

Just renaming the .so wasn't enough, because the soname is stored inside
the so as well and that's what the linked file ends up with regardless
of the actual filename on disk of the .so we link to.

jagerman committed:
Fix android static bundle

jagerman committed:
Minor fixes to build scripts

jagerman committed:
Fix broken option defaults

Moving things around put the option *before* the variable that sets the
default.  Oops.
